src/python/pants/util/frozendict.py

```diff
@@ -0,0 +1,56 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Any, Dict, Iterable, Iterator, Mapping, Optional, Tuple, TypeVar, Union
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+class FrozenDict(Mapping[K, V]):
+    """A wrapper around a normal `dict` that removes all methods to mutate the instance and that
+    implements __hash__.
+
+    This should be used instead of normal dicts when working with the engine because normal dicts
+    are not safe to use.
+    """
+
+    def __init__(self, item: Optional[Union[Mapping[K, V], Iterable[Tuple[K, V]]]] = None) -> None:
+        """Creates a `FrozenDict` from a mapping object or a sequence of tuples representing
+        entries.
+
+        These value should be immutable and hashable, but this collection does not do any validation
+        that that is true.
+        """
+        self._data: Dict[K, V] = dict(item)  # type: ignore[arg-type]
+
+    def __getitem__(self, k: K) -> V:
+        return self._data[k]
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    def __iter__(self) -> Iterator[K]:
+        return iter(self._data)
+
+    def __reversed__(self) -> Iterator[K]:
+        return reversed(tuple(self._data))
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, FrozenDict):
+            return NotImplemented
+        return tuple(self.items()) == tuple(other.items())
+
+    def __hash__(self) -> int:
+        try:
+            return hash(tuple(self.items()))
```

Comment:
I have a hard time believing we construct many FrozenDicts that are never hashes. IOW I'd guess we can do this in the constructor and store the result with marginal perf effects (may even increase perf depending on the average hash call count.

Reply:
> may even increase perf depending on the average hash call count

Rather, we should probably use the lazy-hash pattern Stu added to `FrozenOrderedSet`: 

https://github.com/pantsbuild/pants/blob/eeb9d7488a39bb428bf19e273584675d1041745b/src/python/pants/util/ordered_set.py#L235-L240

We should probably keep the hash calculation lazy - there is a real, albeit minor, cost to calculating the hash.

--

If we still care about validating that the type is hashable in the first place, I think that we could  simply check `hasattr(item, '__hash__')`. That's an extremely light-weight dict lookup on the object's attributes. Almost certainly worth the benefits of eager validation.

Reply:
That's ball fine but my point was, off we ~always calculate the hash we always calculate the hash. The premise very well may be wrong, but given the premise, better to do upfront. I'm fine with experiments and follow up.

Reply:
I went with eager computation of the hash. `hasattr(item, '__hash__')` doesn't work. 

We already know that the keys are hashable because of `dict(item)`, but we don't know that the `values()` are. The `values()` might be heterogenous or be empty, so there's no canonical element of `values()` that we can inspect.

So, the simplest implementation is to eagerly compute the hash, as you suggested. We can make this lazier if we find it to be a performance bottleneck, but for now, I agree with your assessment.

Thanks for the suggestion!